like dislike with user storage

itswwong opened the pull request.

itswwong committed:
working on storing userID when liking/disliking (in progress)

itswwong committed:
add new document in Users during sign up and sign in (in progress)

itswwong committed:
finished adding new documents when signing in or signing up

itswwong committed:
working on updating storing entries when liking/disliking (incomplete))

itswwong committed:
storing users when liking/disliking (in progress))

itswwong committed:
implemented like/dislike with user storage

itswwong committed:
Merge branch 'main' of https://github.com/rainawan/BruinDictionary into like-dislike-users

itswwong committed:
pulled and merged from main

sonyakim-dev commented on src/pages/Dictionary/components/LikeDislikeButtons.jsx:
```diff
@@ -1,67 +1,149 @@
+  const docSnapshot = getDoc(userDocRef);
+
+  useEffect(() => {
+    const checkUserExists = async () => {
```
you dont need to check user exists

sonyakim-dev commented on src/pages/Dictionary/components/LikeDislikeButtons.jsx:
```diff
@@ -1,67 +1,149 @@
+
+      const docSnapshot = await getDoc(userDocRef);
+
+      if (!docSnapshot.exists()) {
```
you dont need to check docSnapshot exists

sonyakim-dev replied:
and you dont need to use getDoc

sonyakim-dev commented on src/pages/Dictionary/components/LikeDislikeButtons.jsx:
```diff
@@ -1,67 +1,149 @@
+  const mutateUser = useFirestoreDocumentMutation(userDocRef, { merge: true });
+  const docSnapshot = getDoc(userDocRef);
+
+  useEffect(() => {
```
no need to use useEffect

sonyakim-dev replied:
this whole useEffect is uneccessary

sonyakim-dev commented on src/pages/Dictionary/components/LikeDislikeButtons.jsx:
```diff
@@ -1,67 +1,149 @@
+          dislikes: {},
+          email: userData.email,
+          likes: {},
+          username: ''
```
just create likes and dislikes only. dont mutate other fields

sonyakim-dev commented on src/pages/SignUp/UserSignUp.jsx:
```diff
@@ -2,6 +2,10 @@ import { useState } from 'react';
 import { useNavigate, Link } from 'react-router-dom';
 import { getAuth, createUserWithEmailAndPassword } from 'firebase/auth';
 import { Button, Input } from '@nextui-org/react';
+import { collection } from 'firebase/firestore';
```
no need to make Users collection when User sign up

sonyakim-dev commented on src/pages/Dictionary/components/LikeDislikeButtons.jsx:
```diff
@@ -1,67 +1,149 @@
+
+  // Early return or handle cases where there is no current user
+  if (!userData) {
+    return <p>Please Sign In</p>;
```
like/dislike button should be allowed to unsigned user too

sonyakim-dev replied:
check if user is defined or not inside the handleAction function

sonyakim-dev commented on src/pages/Dictionary/components/LikeDislikeButtons.jsx:
```diff
@@ -1,67 +1,149 @@
   const handleAction = (newAction) => {
-    // TODO: update firebase with new action
+    let updateEntry = {};
+    let updateUser = {
```
why do we need these updateEntry and updateUser

sonyakim-dev replied:
this is enough
```js
if (userData) {
  mutateUser.mutate({
    likes: { [entry.id]: true },
    dislikes: { [entry.id]: false }
  });
}
```

sonyakim-dev commented on src/utils/getCurrentUserQuery.js:
```diff
@@ -0,0 +1,37 @@
+// import { useEffect, useState } from 'react';
```
remove this file

sonyakim-dev commented on src/utils/fetchData.js:
```diff
@@ -1,5 +1,6 @@
+import { useState, useEffect } from 'react';
 import { useFirestoreQuery } from '@react-query-firebase/firestore';
-import { collection, query } from 'firebase/firestore';
+import { collection, query, doc, getDoc } from 'firebase/firestore';
```
remove unnecessary change

sonyakim-dev commented on src/pages/Dictionary/components/LikeDislikeButtons.jsx:
```diff
@@ -1,67 +1,149 @@
+
+  const mutateEntry = useFirestoreDocumentMutation(entryDocRef, { merge: true });
+  const mutateUser = useFirestoreDocumentMutation(userDocRef, { merge: true });
+  const docSnapshot = getDoc(userDocRef);
```
remove this

sonyakim-dev commented on src/pages/Dictionary/components/LikeDislikeButtons.jsx:
```diff
@@ -1,67 +1,149 @@
-  const entryDocRef = doc(db, 'Entries', entry.id);
-  const mutation = useFirestoreDocumentMutation(entryDocRef, { merge: true });
+  const entryID = entry.id;
+  const userID = userData.userid;
```
this throw error when userData is undefined

sonyakim-dev replied:
you dont need the early return. just doing this does not throw error
```
const entryDocRef = doc(db, 'Entries', entryData.entryid);
const userDocRef = doc(db, 'Users', useData?.userid);
```

sonyakim-dev replied:
and this early return can cause dom rendering issue

itswwong committed:
optimized code, got rid of unnecessary lines/files

itswwong committed:
finished user storage, added loading checks, updated UI

- finished user storage function when liking/disliking
- added loading checks to prevent race conditions.
- buttons now reflects users' previous like/dislike interactions
- like/dislike counters are now fetched from firebase (not changing locally)

itswwong committed:
Merge branch 'main' of https://github.com/rainawan/BruinDictionary into like-dislike-users

itswwong committed:
Merge branch 'main' of https://github.com/rainawan/BruinDictionary into like-dislike-users

itswwong committed:
Merge branch 'main' of https://github.com/rainawan/BruinDictionary into like-dislike-users

itswwong committed:
Merge branch 'main' of https://github.com/rainawan/BruinDictionary into like-dislike-users

itswwong committed:
synced like/dislike display counter with firebase

itswwong committed:
updated to use useFirestoreDocumentMutation
